src/hp/derp/http/client.rs

```diff
@@ -1,39 +1,92 @@
 //! Based on tailscale/derp/derphttp/derphttp_client.go
-use std::net::SocketAddr;
+use std::net::{SocketAddr, TcpStream};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::Arc;
 
 use anyhow::Result;
 use futures::future::BoxFuture;
+use tokio::sync::Mutex;
 
+use crate::hp::derp::{DerpNode, PacketForwarder};
 use crate::hp::key;
 
-use crate::hp::derp::{DerpRegion, ReceivedMessage};
+use crate::hp::derp::{client::Client as DerpClient, DerpRegion, ReceivedMessage};
 
 #[derive(Debug, thiserror::Error, PartialEq, Eq)]
 pub enum ClientError {
     #[error("todo")]
     Todo,
     #[error("closed")]
     Closed,
+    #[error("no derp client")]
+    NoClient,
+    #[error("send")]
+    Send,
 }
 
-#[derive(Default, Debug, Clone, PartialEq, Eq)]
-pub struct Client {}
+/// An HTTP DERP client.
+///
+/// Cheaply clonable.
+#[derive(Clone)]
+pub struct Client {
+    inner: Arc<InnerClient>,
+}
 
-impl Client {
-    pub fn new_region<F>(key: key::node::SecretKey, f: F) -> Self
-    where
-        F: Fn() -> BoxFuture<'static, Option<DerpRegion>>,
-    {
-        // TODO:
-        Client {}
+impl PartialEq for Client {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.inner, &other.inner)
     }
+}
+
+impl Eq for Client {}
+
+impl std::fmt::Debug for Client {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Client {{}}")
+    }
+}
 
-    pub fn set_can_ack_pings(&self, val: bool) {
-        // TODO:
+struct InnerClient {
+    secret_key: key::node::SecretKey,
+    get_region:
+        Option<Box<dyn Fn() -> BoxFuture<'static, Option<DerpRegion>> + Send + Sync + 'static>>,
+    can_ack_pings: bool,
+    is_preferred: Mutex<bool>,
+    derp_client: Mutex<Option<DerpClient<tokio::net::tcp::OwnedReadHalf>>>,
+    is_closed: AtomicBool,
+    address_family_selector:
+        Option<Box<dyn Fn() -> BoxFuture<'static, bool> + Send + Sync + 'static>>,
```

Comment:
the `get_region` and `address_family_selector` fns are pretty ugly to look at, but I think maybe it's just fine?

Reply:
yeah, this is the way to do it unfortunately